File: leetcode/graph/bfs/0752_open_the_lock.py
# LC 752. Open the Lock | Medium
from collections import deque

class Solution:
    def openLock(self, deadends, target):
        dead = set(deadends)
        if '0000' in dead: return -1
        q = deque([('0000', 0)]); visited = {'0000'}
        while q:
            state, turns = q.popleft()
            if state == target: return turns
            for i in range(4):
                d = int(state[i])
                for move in [(d+1)%10, (d-1)%10]:
                    new = state[:i] + str(move) + state[i+1:]
                    if new not in visited and new not in dead:
                        visited.add(new); q.append((new, turns+1))
        return -1

File: leetcode/graph/bfs/0752_open_the_lock.cpp
// LC 752. Open the Lock | Medium
#include <vector>
#include <string>
#include <queue>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> dead(deadends.begin(), deadends.end()), vis;
        if (dead.count("0000")) return -1;
        queue<pair<string,int>> q; q.push({"0000", 0}); vis.insert("0000");
        while (!q.empty()) {
            auto [state, turns] = q.front(); q.pop();
            if (state == target) return turns;
            for (int i = 0; i < 4; i++)
                for (int d : {1, -1}) {
                    string nxt = state; nxt[i] = '0' + (nxt[i]-'0'+d+10)%10;
                    if (!vis.count(nxt) && !dead.count(nxt)) { vis.insert(nxt); q.push({nxt, turns+1}); }
                }
        }
        return -1;
    }
};
